src/main/java/me/gnat008/perworldinventory/process/TeleportProcess.java

```diff
@@ -0,0 +1,295 @@
+package me.gnat008.perworldinventory.process;
+
+import me.gnat008.perworldinventory.PwiLogger;
+import me.gnat008.perworldinventory.config.PwiProperties;
+import me.gnat008.perworldinventory.config.Settings;
+import me.gnat008.perworldinventory.data.players.PWIPlayerManager;
+import me.gnat008.perworldinventory.data.metadata.PWIMetaDataManager;
+import me.gnat008.perworldinventory.groups.Group;
+import me.gnat008.perworldinventory.groups.GroupManager;
+import me.gnat008.perworldinventory.permission.PermissionManager;
+import me.gnat008.perworldinventory.permission.PlayerPermission;
+import org.bukkit.GameMode;
+import org.bukkit.entity.Player;
+import org.bukkit.Location;
+import org.bukkit.Bukkit;
+import org.bukkit.Server;
+import org.bukkit.World;
+import org.bukkit.event.player.PlayerTeleportEvent;
+
+import java.util.Map;
+
+import javax.inject.Inject;
+
+/**
+ * Process to follow when a player's inventory is changing.
+ */
+public class TeleportProcess {
+
+    @Inject
+    private GroupManager groupManager;
+
+    @Inject
+    private PermissionManager permissionManager;
+
+    @Inject
+    private PWIPlayerManager playerManager;
+
+    @Inject
+    private Settings settings;
+
+    @Inject
+    private PWIMetaDataManager metaDataManager;
+
+    @Inject
+    private Server server;
+
+    TeleportProcess() {
+    }
+
+    /**
+     * Begins the process of a player teleporting.
+     *
+     * @param event The called {@link PlayerTeleportEvent}.
+     */
+    public void preProcessTeleport(PlayerTeleportEvent event) {
+        Player player = event.getPlayer();
+        Location from = event.getFrom();
+        Location to = event.getTo();
+
+        if(event.isCancelled()) {
+            PwiLogger.debug(String.format("Player '%s' tried to teleport from '%s' to '%s', but the event was previously cancelled, dismissing handler",
+              player.getName(),
+              from,
+              to));
+            return;
+        }
+
+        if(from == null || from.getWorld() == null) {
+            PwiLogger.warning(String.format("Player '%s' tried to teleport from '%s' to '%s', but got null on origin. Cancelling teleport.",
+              player.getName(),
+              from,
+              to));
+            event.setCancelled(true);
+            return;
+        }
+
+        if(to == null || to.getWorld() == null) {
+            PwiLogger.warning(String.format("Player '%s' tried to teleport from '%s' to '%s', but got null on origin. Cancelling teleport.",
+              player.getName(),
+              from,
+              to));
+            event.setCancelled(true);
+            return;
+        }
+
+        if(from.getWorld().equals(to.getWorld())) {
+            PwiLogger.debug(String.format("Player '%s' is teleporting from world '%s' to world '%s', in-world teleport, dismissing handler.",
+              player.getName(),
+              from.getWorld().getName(),
+              to.getWorld().getName()));
+            return;
```

Comment:
That would disable the regular /tp-command.

The reason I was canceling the teleport earlier is because it was important for the rest of the code to know which world you just were in, in order to save that location.

I'm actually considering refactoring the `OnWorldChange` back, so I can be more lax with how teleport events are executed, as I thought I'd merge the two functions together when I refactored it into the teleport event, which obviously didn't happen.

That way, I can let teleports slip trough that are uncertain, since the position is less important than the inventory.